[pkg/ottl] Add client metadata access to OTTL contexts

ronkorland opened the pull request:
#### Description

This PR adds metadata access capability to all OTTL (OpenTelemetry Transformation Language) contexts, enabling users to access client request metadata within OTTL transformations.

**What this achieves:**
- Adds new `metadata` path expression to access the entire client metadata object
- Adds `metadata["key"]` path expression to access specific metadata values by key
- Available in all OTTL contexts (span, log, metric, resource, datapoint, scope, spanevent, profile, profilesample)
- Enables access to HTTP headers, authentication tokens, and other client metadata from incoming requests
- Supports filtering, routing, and enrichment use cases based on request metadata

**Implementation details:**
- Created new `ctxmetadata` package with path expression parser
- Integrated metadata access into the common OTTL context parser
- Returns `client.Metadata` type for full metadata access
- Returns `string` or `nil` for specific key access (first value if multiple exist)
- Read-only access (metadata cannot be modified via OTTL)

#### Link to tracking issue
Fixes #33288

#### Testing

- Added comprehensive unit tests in `pkg/ottl/contexts/internal/ctxmetadata/metadata_test.go`
- Tests cover accessing entire metadata object and specific keys
- Tests validate behavior with empty keys, non-existent keys, and multi-value scenarios
- Tests verify error handling for invalid access patterns
- All existing OTTL context tests continue to pass

#### Documentation

- Updated README files for all OTTL contexts to document the new metadata paths:
  - `pkg/ottl/contexts/ottldatapoint/README.md`
  - `pkg/ottl/contexts/ottllog/README.md` 
  - `pkg/ottl/contexts/ottlmetric/README.md`
  - `pkg/ottl/contexts/ottlprofile/README.md`
  - `pkg/ottl/contexts/ottlprofilesample/README.md`
  - `pkg/ottl/contexts/ottlresource/README.md`
  - `pkg/ottl/contexts/ottlscope/README.md`
  - `pkg/ottl/contexts/ottlspan/README.md`
  - `pkg/ottl/contexts/ottlspanevent/README.md`
- Each README now includes:
  - `metadata` path description and return type (`client.Metadata`)
  - `metadata[""]` path description and return type (`string` or `nil`)
  - Clear explanation of behavior for missing keys and multiple values

Commit:
[pkg/ottl] Add client metadata access to OTTL contexts

Add `metadata` and `metadata["key"]` path expressions to access
client request metadata in all OTTL transformation contexts.

Fixes #33288

Commit:
Merge branch 'main' into metadata-context

# Conflicts:
#	pkg/ottl/go.mod
#	pkg/ottl/go.sum

Commit:
fix import into the right group

ronkorland commented on pkg/ottl/contexts/internal/ctxcommon/parser.go:
```diff
@@ -4,6 +4,7 @@
 package ctxcommon // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/internal/ctxcommon"
 
 import (
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/internal/ctxmetadata"
```
fixed

Commit:
fix missing new collector/client dependency

Commit:
feat(ottl/context): add namespaced context paths; migrate metadata to context.client.metadata; update docs and tests

- Add internal/ctxcontext README documenting:
  - context.client.metadata, context.client.metadata[""]
  - context.client.addr
  - context.client.auth.attributes, context.client.auth.attributes[""]
  - context.grpc.metadata, context.grpc.metadata[""]
- Update READMEs for all OTTL contexts (log/span/metric/spanevent/resource/scope/datapoint/profile/profilesample) to use context.client.metadata and include the new context.* paths.
- Update ctxcommon parser to route context.* paths via ctxcontext and remove deprecated ctxmetadata import.
- Expand ctxcontext tests:
  - Cover client metadata (entire/keys/empty/missing/multi-values), client addr, client auth attributes (map/key/missing), and grpc metadata (entire/keys/missing).
  - Make all paths read-only assertions.
  - Improve test function names and fix local variable collision with metadata package.

ronkorland commented:
Hi @edmocosta, thanks for you feedback
I change in my last commit according to your recommendation
Please review and let me know if something is missing

Commit:
fix lint errors + fix go.mod

Commit:
Merge branch 'main' into metadata-context

Commit:
fix lint

Commit:
Merge branch 'main' into metadata-context

# Conflicts:
#	connector/countconnector/go.mod
#	connector/countconnector/go.sum
#	connector/signaltometricsconnector/go.mod
#	connector/signaltometricsconnector/go.sum
#	connector/sumconnector/go.mod
#	connector/sumconnector/go.sum
#	internal/filter/go.mod
#	internal/filter/go.sum
#	pkg/ottl/go.mod
#	pkg/ottl/go.sum
#	processor/filterprocessor/go.mod
#	processor/filterprocessor/go.sum
#	processor/logdedupprocessor/go.mod
#	processor/logdedupprocessor/go.sum
#	processor/spanprocessor/go.mod
#	processor/spanprocessor/go.sum
#	processor/tailsamplingprocessor/go.mod
#	processor/tailsamplingprocessor/go.sum
#	processor/transformprocessor/go.mod
#	processor/transformprocessor/go.sum

axw reviewed (commented):
Really glad to see this, thank you @ronkorland :pray:

Just a comment for future consideration, not proposing we change anything here: I think it would be ideal if OTTL paths could be injected into the context, which would enable receivers to introduce their own domain-specific context.

For example, it would be ideal if kafkareceiver could expose the Kafka topic, partition, and message key as a metadata path. See also https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/35231

That way pkg/ottl could be loosely coupled with the data source.

axw commented:
@ronkorland the problem is that the Kafka receiver already adds message headers to as client metadata, and if we were to add message key, topic, etc. then they may conflict with message headers. If you are planning to work on that, let's discuss on https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/35231 first please.

I'll butt out of this PR now, I don't want to derail :)

edmocosta reviewed (commented):
Thank you for working on this @ronkorland, I've left a few suggestions for your consideration. I'll continue reviewing/testing it after we close them, as I think it would change the code a bit.

Commit:
Merge branch 'main' into metadata-context

# Conflicts:
#	pkg/ottl/go.mod

Commit:
change return type to `pcommon.Slice` and `pcommon.Map` and change function names

Commit:
update docs + add `ottlcontext` + move `context` from parser to contexts

Commit:
Merge branch 'main' into metadata-context

ronkorland commented:
Thanks for the review, @edmocosta 
I’ve addressed the suggestions you left. Could you please continue reviewing/testing the updated code when you get a chance?

Commit:
Merge branch 'main' into metadata-context

# Conflicts:
#	connector/countconnector/go.mod
#	connector/countconnector/go.sum
#	connector/signaltometricsconnector/go.mod
#	connector/signaltometricsconnector/go.sum
#	connector/sumconnector/go.mod
#	connector/sumconnector/go.sum
#	internal/filter/go.mod
#	internal/filter/go.sum
#	pkg/ottl/go.mod
#	pkg/ottl/go.sum
#	processor/filterprocessor/go.mod
#	processor/filterprocessor/go.sum
#	processor/logdedupprocessor/go.mod
#	processor/logdedupprocessor/go.sum
#	processor/spanprocessor/go.mod
#	processor/spanprocessor/go.sum
#	processor/tailsamplingprocessor/go.mod
#	processor/tailsamplingprocessor/go.sum
#	processor/transformprocessor/go.mod
#	processor/transformprocessor/go.sum

Commit:
fix gotidy after merge with main

Commit:
fix lint

ronkorland commented:
@edmocosta, I fixed the merge conflicts and the lint errors

Commit:
Merge branch 'main' into metadata-context

Commit:
Merge branch 'main' into metadata-context

# Conflicts:
#	connector/countconnector/go.mod
#	connector/countconnector/go.sum
#	connector/signaltometricsconnector/go.mod
#	connector/signaltometricsconnector/go.sum
#	connector/sumconnector/go.mod
#	connector/sumconnector/go.sum
#	internal/filter/go.mod
#	internal/filter/go.sum
#	pkg/ottl/go.mod
#	pkg/ottl/go.sum
#	processor/filterprocessor/go.mod
#	processor/filterprocessor/go.sum
#	processor/logdedupprocessor/go.mod
#	processor/logdedupprocessor/go.sum
#	processor/spanprocessor/go.mod
#	processor/spanprocessor/go.sum
#	processor/tailsamplingprocessor/go.mod
#	processor/tailsamplingprocessor/go.sum
#	processor/transformprocessor/go.mod
#	processor/transformprocessor/go.sum

edmocosta commented:
Hi @ronkorland, I'm sorry for the delay, there's nothing missing from your side. I was on PTO + very busy weeks, but I will do it ASAP. Thank you!

edmocosta commented on pkg/ottl/contexts/internal/ctxcontext/context.go:
```diff
@@ -0,0 +1,306 @@
+		return authMap, nil
+	}
+	names := authData.GetAttributeNames()
+	for _, name := range names {
+		attrVal := authData.GetAttribute(name)
+		attrStr, err := getAuthAttributeValue(attrVal)
+		if err != nil {
+			return pcommon.NewMap(), err
+		}
+		switch v := attrStr.(type) {
+		case string:
+			authMap.PutStr(name, v)
+		default:
+			authMap.PutEmpty(name)
+		}
+	}
+	return authMap, nil
```
```suggestion
	for _, name := range names {
		attrVal := authData.GetAttribute(name)
		authMapAttrVal := authMap.PutEmpty(name)
		switch typedAttrVal := attrVal.(type) {
		case string:
			authMapAttrVal.SetStr(typedAttrVal)
		case []string:
			slice := authMapAttrVal.SetEmptySlice()
			slice.EnsureCapacity(len(typedAttrVal))
			for _, str := range typedAttrVal {
				slice.AppendEmpty().SetStr(str)
			}
		default:
			_ = authMapAttrVal.FromRaw(attrVal)
		}
	}
	return authMap, nil
```

According to the `AuthData.GetAttribute`, it normally returns a `string`, and in a few cases`[]string`.  I got the JSON marshalling fallback idea, but IMO, we shouldn't handle that yet, at least not on the first version. 
If for some reason the data the `GetAttribute` returns is not supported by OTTL/PData, we can keep it empty, and handle those special cases one-by-one as they appear. WDYT?

edmocosta commented on pkg/ottl/contexts/internal/ctxcontext/context.go:
```diff
@@ -0,0 +1,306 @@
+}
+
+func convertClientMetadataToMap(md client.Metadata) pcommon.Map {
+	mdMap := pcommon.NewMap()
+	for k := range md.Keys() {
+		sl := mdMap.PutEmptySlice(k)
+		mdVal := md.Get(k)
+		mdSl := convertStringArrToValueSlice(mdVal)
+		mdSl.Slice().CopyTo(sl)
+	}
+	return mdMap
```
```suggestion
	mdMap := pcommon.NewMap()
	for k := range md.Keys() {
		convertStringArrToValueSlice(md.Get(k)).MoveTo(mdMap.PutEmpty(k))
	}
	return mdMap
```

Let's `MoveTo` instead of `CopyTo`, as the value source is only used for this function .

ronkorland replied:
done

edmocosta commented on pkg/ottl/contexts/internal/ctxcontext/context.go:
```diff
@@ -0,0 +1,306 @@
+			if len(mdVal) == 0 {
+				return nil, nil
+			}
+			return getIndexableValueFromStringArr(ctx, tCtx, keys, mdVal)
```
```suggestion
			return getIndexableValueFromStringArr(ctx, tCtx, keys[1:], mdVal)
```

Same as my previous comment.

edmocosta commented on pkg/ottl/contexts/internal/ctxcontext/context.go:
```diff
@@ -0,0 +1,306 @@
+			return cl.Addr.String(), nil
+		},
+		Setter: func(_ context.Context, _ K, _ any) error {
+			return errors.New("cannot set value in context.client.addr")
```
```suggestion
			return errors.New("context.client.addr is read-only and cannot be modified")
```

Just to make it clear why the set failed. If you agree, could you please also apply it to all other messages?

edmocosta replied:
I saw you applied this suggestion only to this setter, could you please do the same for all others paths? We can maybe create a const for that message so we don't have to repeat it multiple times: 

```go
const readOnlyPathErrMsg = "%q is read-only and cannot be modified"
return fmt.Errorf(readOnlyPathErrMsg, "context.client.addr")
```

edmocosta commented on pkg/ottl/contexts/ottldatapoint/README.md:
```diff
@@ -49,6 +49,13 @@ The following paths are supported.
 | datapoint.scale                                | the scale of the data point being processed                                                                                                                                         | int64                                                                   |
 | datapoint.zero_count                           | the zero_count of the data point being processed                                                                                                                                    | int64                                                                   |
 | datapoint.quantile_values                      | the quantile_values of the data point being processed                                                                                                                               | pmetric.SummaryDataPointValueAtQuantileSlice                            |
+| context.client.addr                            | the remote address string from the client info                                                                                                                                      | string                                                                  |
```
We can add these as link to the `context` docs, so we don't need to keep it in sync? similar to the `span.*` on https://github.com/ronkorland/opentelemetry-collector-contrib/blob/main/pkg/ottl/contexts/ottlspanevent/README.md

ronkorland committed:
Merge branch 'main' into metadata-context

# Conflicts:
#	connector/countconnector/go.mod
#	connector/countconnector/go.sum
#	connector/signaltometricsconnector/go.mod
#	connector/signaltometricsconnector/go.sum
#	connector/sumconnector/go.mod
#	connector/sumconnector/go.sum
#	internal/filter/go.mod
#	internal/filter/go.sum
#	pkg/ottl/contexts/ottlprofile/README.md
#	pkg/ottl/contexts/ottlprofilesample/README.md
#	pkg/ottl/go.mod
#	pkg/ottl/go.sum
#	processor/filterprocessor/go.mod
#	processor/filterprocessor/go.sum
#	processor/logdedupprocessor/go.mod
#	processor/logdedupprocessor/go.sum
#	processor/spanprocessor/go.mod
#	processor/spanprocessor/go.sum
#	processor/tailsamplingprocessor/go.mod
#	processor/tailsamplingprocessor/go.sum
#	processor/transformprocessor/go.mod
#	processor/transformprocessor/go.sum

ronkorland committed:
make gotidy

ronkorland committed:
CR fixes

ronkorland committed:
fix README files

ronkorland committed:
Merge branch 'main' into metadata-context

# Conflicts:
#	connector/countconnector/go.mod
#	connector/countconnector/go.sum
#	connector/signaltometricsconnector/go.mod
#	connector/signaltometricsconnector/go.sum
#	connector/sumconnector/go.mod
#	connector/sumconnector/go.sum
#	internal/filter/go.mod
#	internal/filter/go.sum
#	pkg/ottl/go.mod
#	pkg/ottl/go.sum
#	processor/filterprocessor/go.mod
#	processor/filterprocessor/go.sum
#	processor/logdedupprocessor/go.mod
#	processor/logdedupprocessor/go.sum
#	processor/spanprocessor/go.mod
#	processor/spanprocessor/go.sum
#	processor/tailsamplingprocessor/go.mod
#	processor/tailsamplingprocessor/go.sum
#	processor/transformprocessor/go.mod
#	processor/transformprocessor/go.sum

ronkorland committed:
fix CR + tests

edmocosta reviewed (commented):
Thank you @ronkorland for working on it, it looks good so far!
There's a CLA issue with your latest commits, which seems to be from another GH account. Could you please take a look at that while I run a few final tests? Thanks!

ronkorland committed:
Merge branch 'main' into metadata-context

ronkorland committed:
revert getIndexableValue

ronkorland committed:
CR

ronkorland committed:
fix test and add ` around context.

ronkorland committed:
fix lint error

edmocosta commented on pkg/ottl/contexts/ottlcontext/README.md:
```diff
@@ -0,0 +1,21 @@
+|------------------------------------|------------------------------------------------------------------------------------------------------------|----------------------|
+| context.client.addr                | the remote address string from the client info                                                             | string               |
+| context.client.metadata            | client metadata attached to the request via `go.opentelemetry.io/collector/client`                         | pcommon.Map          |
+| context.client.metadata[""]        | the value for a specific metadata key. Returns nil if key missing                                          | pcommon.Slice or nil |
```
```suggestion
| context.client.metadata[""]        | the value for a specific metadata key.                                           | pcommon.Slice, string or nil |
```

edmocosta reviewed (commented):
Thank you @ronkorland, I ran a bunch of tests and the only issue I've found was accessing `client.auth.attributes["roles"]` with sub-indexes, for example `client.auth.attributes["roles"][0]`. All other use-cases I could think of are working pretty well, great work! I've added a couple of suggestion for fixing that issue, plus some docs suggestions to reflect the latest changes. After that, I think we're good!  

@evan-bradley @TylerHelmuth please have a look once you a get a chance.

ronkorland committed:
Merge branch 'main' into metadata-context

# Conflicts:
#	connector/countconnector/go.mod
#	connector/countconnector/go.sum
#	connector/signaltometricsconnector/go.mod
#	connector/signaltometricsconnector/go.sum
#	connector/sumconnector/go.mod
#	connector/sumconnector/go.sum
#	internal/filter/go.mod
#	internal/filter/go.sum
#	pkg/ottl/contexts/ottlprofile/README.md
#	pkg/ottl/go.mod
#	pkg/ottl/go.sum
#	processor/filterprocessor/go.mod
#	processor/filterprocessor/go.sum
#	processor/logdedupprocessor/go.mod
#	processor/logdedupprocessor/go.sum
#	processor/spanprocessor/go.mod
#	processor/spanprocessor/go.sum
#	processor/tailsamplingprocessor/go.mod
#	processor/tailsamplingprocessor/go.sum
#	processor/transformprocessor/go.mod
#	processor/transformprocessor/go.sum

ronkorland committed:
Update pkg/ottl/contexts/ottlcontext/README.md

Co-authored-by: Edmo Vamerlatti Costa <11836452+edmocosta@users.noreply.github.com>

ronkorland committed:
remove comment

ronkorland committed:
Merge branch 'metadata-context' of github.com:ronkorland/opentelemetry-collector-contrib into metadata-context

ronkorland committed:
Update pkg/ottl/contexts/internal/ctxcontext/context_test.go

Co-authored-by: Edmo Vamerlatti Costa <11836452+edmocosta@users.noreply.github.com>

ronkorland committed:
Update pkg/ottl/contexts/internal/ctxcontext/context.go

Co-authored-by: Edmo Vamerlatti Costa <11836452+edmocosta@users.noreply.github.com>

ronkorland committed:
Update pkg/ottl/contexts/internal/ctxcontext/context.go

Co-authored-by: Edmo Vamerlatti Costa <11836452+edmocosta@users.noreply.github.com>

ronkorland committed:
Update pkg/ottl/contexts/internal/ctxcontext/context.go

Co-authored-by: Edmo Vamerlatti Costa <11836452+edmocosta@users.noreply.github.com>

ronkorland committed:
Update pkg/ottl/contexts/internal/ctxcontext/context.go

Co-authored-by: Edmo Vamerlatti Costa <11836452+edmocosta@users.noreply.github.com>

ronkorland committed:
Merge branch 'metadata-context' of github.com:ronkorland/opentelemetry-collector-contrib into metadata-context

ronkorland committed:
fix imports

ronkorland committed:
Merge branch 'main' into metadata-context

# Conflicts:
#	connector/countconnector/go.mod
#	connector/countconnector/go.sum
#	connector/signaltometricsconnector/go.mod
#	connector/signaltometricsconnector/go.sum
#	connector/sumconnector/go.mod
#	connector/sumconnector/go.sum
#	internal/filter/go.mod
#	internal/filter/go.sum
#	pkg/ottl/go.mod
#	pkg/ottl/go.sum
#	processor/filterprocessor/go.mod
#	processor/filterprocessor/go.sum
#	processor/logdedupprocessor/go.mod
#	processor/logdedupprocessor/go.sum
#	processor/spanprocessor/go.mod
#	processor/spanprocessor/go.sum
#	processor/tailsamplingprocessor/go.mod
#	processor/tailsamplingprocessor/go.sum
#	processor/transformprocessor/go.mod
#	processor/transformprocessor/go.sum

ronkorland committed:
Merge branch 'main' into metadata-context

# Conflicts:
#	connector/countconnector/go.mod
#	connector/countconnector/go.sum
#	connector/signaltometricsconnector/go.mod
#	connector/signaltometricsconnector/go.sum
#	connector/sumconnector/go.mod
#	connector/sumconnector/go.sum
#	internal/filter/go.mod
#	internal/filter/go.sum
#	pkg/ottl/go.mod
#	pkg/ottl/go.sum
#	processor/attributesprocessor/go.mod
#	processor/filterprocessor/go.mod
#	processor/filterprocessor/go.sum
#	processor/logdedupprocessor/go.mod
#	processor/logdedupprocessor/go.sum
#	processor/spanprocessor/go.mod
#	processor/spanprocessor/go.sum
#	processor/tailsamplingprocessor/go.mod
#	processor/tailsamplingprocessor/go.sum
#	processor/transformprocessor/go.mod
#	processor/transformprocessor/go.sum

edmocosta commented:
Thank you @ronkorland, it looks good to me! Unfortunately it has a few conflicts to resolve (again).
@TylerHelmuth @evan-bradley please let me know your thoughts, otherwise we're good to merge once CI is green.

ronkorland committed:
Merge branch 'main' into metadata-context

# Conflicts:
#	connector/countconnector/go.mod
#	connector/countconnector/go.sum
#	connector/signaltometricsconnector/go.mod
#	connector/signaltometricsconnector/go.sum
#	connector/sumconnector/go.mod
#	connector/sumconnector/go.sum
#	internal/filter/go.mod
#	internal/filter/go.sum
#	pkg/ottl/go.mod
#	pkg/ottl/go.sum
#	processor/filterprocessor/go.mod
#	processor/filterprocessor/go.sum
#	processor/logdedupprocessor/go.mod
#	processor/logdedupprocessor/go.sum
#	processor/spanprocessor/go.mod
#	processor/spanprocessor/go.sum
#	processor/tailsamplingprocessor/go.mod
#	processor/tailsamplingprocessor/go.sum
#	processor/transformprocessor/go.mod
#	processor/transformprocessor/go.sum

evan-bradley commented on pkg/ottl/contexts/internal/ctxcontext/context.go:
```diff
@@ -0,0 +1,328 @@
+
+const (
+	readOnlyPathErrMsg = "%q is read-only and cannot be modified"
+	Name               = "context"
```
I know I'm coming into this very late, but I want to brainstorm other names since I feel like `context` doesn't clearly indicate what we're working with to users who aren't familiar with the technical details of how Go's `context.Context` object works.

So far, we're working with a request-scoped context, such as a network client or data on top of a network client like gRPC metadata. It's possible we could extend this to work with things like scraping receivers and would want to name the OTTL context something like `transmission` to indicate that it's data associated with a (pdata) payload going through the pipeline, but I'm a little reluctant to make the name more generic if we don't have a use case in mind for it. What do you think?

edmocosta replied:
Thank you @evan-bradley for sharing your thoughts!

Naming is hard. I wasn’t pretty sure about `context` as well, but ended up suggesting it [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/33288#issuecomment-3151303524) considering the different data sources it can have, and I think this context idea is somehow similar to https://opentelemetry.io/docs/specs/otel/context, or not, IDK :)  

> I'm a little reluctant to make the name more generic if we don't have a use case in mind for it. What do you think?

I think Kafka, and possibly other pull-based receivers, are examples of that (https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/35231). Kafka seems to be passing the Kafka message's headers on the client info, which IMO doesn't fit that well into the request-scoped context (although it might be). 
The collector lib calls it `client`, and we could possibly use that same term, but I'm honestly not sure what's the best option here, but I'd would go with the more generic one, considering there are more components interested on this feature, and they might storing "any" kind of data on it.

evan-bradley commented:
@ronkorland just want to give you an update: we had an offline discussion today and are still thinking through the best name for this new context. If you have time and are willing, feel free to fix the existing merge conflicts and other comments on the PR and we can come back to the name afterward. Otherwise we should hopefully have an update for you shortly.

Thank you for your continued patience, I know this PR has been open for very long now.

ronkorland commented:
@evan-bradley, any updates on the naming discussion?

axw commented:
Is there a conclusion on the naming? Is `otelcol` it?

FWIW, I think that `otelcol` doesn't tell much about what it means. I agree that `context` is too overloaded. `external` feels too restrictive, since you could for example inject headers/metadata within a pipeline.

My 2c: I think `metadata` as originally proposed is appropriate. Metadata is "data about data", which seems totally apt here. One thing I would note is that `metadata["..."]` doesn't leave room for accessing metadata other than request/record headers, e.g. client IP, auth data, Kafka topic/partition, etc. You could shove all of those things in metadata, but that may end up mixing different metadata sources (e.g. Kafka record headers + topic name/partition). So one more suggestion: `metadata.headers["..."]`.

Anyway, I'd rather see the functionality in, so if code owners all think otelcol is best then I would disagree and commit.

edmocosta commented:
Thanks @axw, naming is hard!

> Is there a conclusion on the naming? Is otelcol it?

Yes, we've agreed on `otelcol`, given this context will be used to access existing and future collector's `context.Context` data.

> My 2c: I think metadata as originally proposed is appropriate. Metadata is "data about data", which seems totally apt here. One thing I would note is that metadata["..."] doesn't leave room for accessing metadata other than request/record headers, e.g. client IP, auth data, Kafka topic/partition, etc. You could shove all of those things in metadata, but that may end up mixing different metadata sources (e.g. Kafka record headers + topic name/partition). So one more suggestion: metadata.headers["..."].

Ideally, we need a common prefix for this context, so we can expose all values/paths under that prefix, without having to add multiple OTTL contexts for that. The name must be "generic" and avoid conflicts with existing and future paths. For example, `metadata` is also a `metric` context's field, and if we prefix this one like that, we probably would have issues inferring the context from non-prefixed statements.

Finally, we also want to make it clear that this context data is not pdata, so users know they're accessing/modifying values that are only available inside the OTel collector.

Thank you!

edmocosta committed:
Merge branch 'main' of https://github.com/open-telemetry/opentelemetry-collector-contrib into metadata-context

# Conflicts:
#	connector/countconnector/go.mod
#	connector/countconnector/go.sum
#	connector/signaltometricsconnector/go.mod
#	connector/signaltometricsconnector/go.sum
#	connector/sumconnector/go.mod
#	connector/sumconnector/go.sum
#	internal/filter/go.mod
#	internal/filter/go.sum
#	pkg/ottl/contexts/ottldatapoint/datapoint.go
#	pkg/ottl/contexts/ottllog/log.go
#	pkg/ottl/contexts/ottlmetric/metrics.go
#	pkg/ottl/contexts/ottlprofile/profile.go
#	pkg/ottl/contexts/ottlprofilesample/profilesample.go
#	pkg/ottl/contexts/ottlresource/resource.go
#	pkg/ottl/contexts/ottlscope/scope.go
#	pkg/ottl/contexts/ottlspan/README.md
#	pkg/ottl/contexts/ottlspan/span.go
#	pkg/ottl/go.mod
#	pkg/ottl/go.sum
#	processor/attributesprocessor/go.mod
#	processor/attributesprocessor/go.sum
#	processor/filterprocessor/go.mod
#	processor/filterprocessor/go.sum
#	processor/logdedupprocessor/go.mod
#	processor/logdedupprocessor/go.sum
#	processor/spanprocessor/go.mod
#	processor/spanprocessor/go.sum
#	processor/tailsamplingprocessor/go.mod
#	processor/tailsamplingprocessor/go.sum
#	processor/transformprocessor/go.mod
#	processor/transformprocessor/go.sum

edmocosta committed:
Renaming from `context` to `otelcol`

axw commented:
@ronkorland are you planning to come back to this?

edmocosta committed:
Merge branch 'main' into metadata-context

edmocosta committed:
Go mod tidy

edmocosta commented on pkg/ottl/contexts/ottlcontext/README.md:
```diff
@@ -0,0 +1,22 @@
+# Context (OTTL)
+
+The Context exposes request-related data from the Go `context.Context` passed through the Collector, such as client information and incoming gRPC metadata. Use these paths via the top-level `context` segment, optionally followed by sub-segments.
```
This is stale now after the renaming & discussions, could you please review it again?

edmocosta committed:
make goporto

edmocosta committed:
Update pkg/ottl/contexts/internal/ctxotelcol/otelcol_test.go

Co-authored-by: Evan Bradley <11745660+evan-bradley@users.noreply.github.com>

edmocosta committed:
Merge branch 'main' of https://github.com/open-telemetry/opentelemetry-collector-contrib into metadata-context

# Conflicts:
#	connector/countconnector/go.mod
#	connector/signaltometricsconnector/go.mod
#	connector/signaltometricsconnector/go.sum
#	connector/sumconnector/go.mod
#	connector/sumconnector/go.sum
#	internal/filter/go.mod
#	internal/filter/go.sum
#	pkg/ottl/go.mod
#	pkg/ottl/go.sum
#	processor/filterprocessor/go.mod
#	processor/filterprocessor/go.sum
#	processor/logdedupprocessor/go.mod
#	processor/logdedupprocessor/go.sum
#	processor/spanprocessor/go.mod
#	processor/spanprocessor/go.sum
#	processor/tailsamplingprocessor/go.mod
#	processor/tailsamplingprocessor/go.sum
#	processor/transformprocessor/go.mod
#	processor/transformprocessor/go.sum

edmocosta committed:
Change otelcol TransformContext to use pointer and update docs to include security tips

edmocosta committed:
Add otelcol feature gate

edmocosta committed:
Merge branch 'main' into metadata-context

edmocosta committed:
make gotidy

edmocosta committed:
Merge branch 'main' into metadata-context

edmocosta committed:
Merge branch 'main' of https://github.com/open-telemetry/opentelemetry-collector-contrib into metadata-context

# Conflicts:
#	connector/countconnector/go.mod
#	connector/signaltometricsconnector/go.mod
#	connector/signaltometricsconnector/go.sum
#	connector/sumconnector/go.mod
#	connector/sumconnector/go.sum
#	internal/filter/go.mod
#	internal/filter/go.sum
#	pkg/ottl/go.mod
#	pkg/ottl/go.sum
#	processor/filterprocessor/go.mod
#	processor/filterprocessor/go.sum
#	processor/logdedupprocessor/go.mod
#	processor/logdedupprocessor/go.sum
#	processor/lookupprocessor/go.mod
#	processor/lookupprocessor/go.sum
#	processor/spanprocessor/go.mod
#	processor/spanprocessor/go.sum
#	processor/tailsamplingprocessor/go.mod
#	processor/tailsamplingprocessor/go.sum
#	processor/transformprocessor/go.mod
#	processor/transformprocessor/go.sum

edmocosta committed:
make gotidy

edmocosta committed:
Fix setter errors context name

edmocosta commented:
/easycla

edmocosta committed:
Fix context name, comments and error format

edmocosta committed:
Merge branch 'main' of https://github.com/open-telemetry/opentelemetry-collector-contrib into metadata-context

# Conflicts:
#	connector/countconnector/go.mod
#	connector/signaltometricsconnector/go.mod
#	connector/signaltometricsconnector/go.sum
#	connector/sumconnector/go.mod
#	connector/sumconnector/go.sum
#	internal/filter/go.mod
#	internal/filter/go.sum
#	pkg/ottl/go.mod
#	pkg/ottl/go.sum
#	processor/filterprocessor/go.mod
#	processor/filterprocessor/go.sum
#	processor/logdedupprocessor/go.mod
#	processor/logdedupprocessor/go.sum
#	processor/lookupprocessor/go.mod
#	processor/lookupprocessor/go.sum
#	processor/spanprocessor/go.mod
#	processor/spanprocessor/go.sum
#	processor/tailsamplingprocessor/go.mod
#	processor/tailsamplingprocessor/go.sum
#	processor/transformprocessor/go.mod
#	processor/transformprocessor/go.sum

evan-bradley reviewed (approved):
Thanks @ronkorland. Looks good to me.

ronkorland committed:
Enhance `otelcol` context feature gate registration with additional metadata

- Added a reference URL for issue tracking related to the `otelcol` context.
- Specified the version from which the feature gate is available.

ronkorland committed:
Refactor `otelcol` context handling by removing unused functions and simplifying code

- Removed several unused functions related to gRPC and client metadata accessors.
- Consolidated logic for converting string arrays to value slices.
- Improved code readability and maintainability by eliminating redundant error handling and comments.

ronkorland committed:
fix commit

ronkorland committed:
Merge main into metadata-context
Resolved go.mod/go.sum conflicts by taking main's versions; re-added collector/client in pkg/ottl for otelcol context.

ronkorland commented:
@evan-bradley I've addressed your feedback:
Feature Gate: Added the reference URL and version (v0.147.0).
Refactoring: Cleaned up the `otelcol` path handling and context logic.

evan-bradley commented:
Thanks @ronkorland. Looks good to me, just needs a `make gotidy`.

ronkorland committed:
run gotidy

ronkorland commented:
> Thanks @ronkorland. Looks good to me, just needs a `make gotidy`.

done
